lib/seattleflu/id3c/cli/command/etl/redcap_det_clinical_data_pulls.py

```diff
@@ -0,0 +1,280 @@
+
+    specimen_resource = create_specimen_resource(
+        [specimen_identifier], patient_reference, specimen_type
+    )
+
+    return create_entry_and_reference(specimen_resource, "Specimen")
+
+
+def create_resident_locations(db: DatabaseSession, cache: TTLCache, record: dict) -> Optional[tuple]:
+    """
+    Returns FHIR Location resource entry and reference for resident address
+    and Location resource entry for Census tract.
+    """
+    if not record["address"]:
+        LOG.debug("No address found in REDCap record")
+        return None, None
+
+    address = {
+        "street" : record["address"],
+        "secondary": None,
+        "city": None,
+        "state": None,
+        "zipcode": None
+    }
+
+    lat, lng, canonicalized_address = get_response_from_cache_or_geocoding(address, cache)
+
+    if not canonicalized_address:
+        LOG.debug("Geocoding of address failed")
+        return None, None
+
+    location_type_system = 'http://terminology.hl7.org/CodeSystem/v3-RoleCode'
+    location_type = create_codeable_concept(location_type_system, 'PTRES')
+    location_entries: List[dict] = []
+    location_references: List[dict] = []
+    address_partOf: Dict = None
+
+    tract = location_lookup(db, (lat,lng), 'tract')
+
+    if tract and tract.identifier:
+        tract_identifier = create_identifier(f"{SFS}/location/tract", tract.identifier)
+        tract_location = create_location_resource([location_type], [tract_identifier])
+        tract_entry, tract_reference = create_entry_and_reference(tract_location, "Location")
+        # tract_reference is not used outside of address_partOf so does not
+        # not need to be appended to the list of location_references.
+        address_partOf = tract_reference
+        location_entries.append(tract_entry)
+
+    address_hash = generate_hash(canonicalized_address)
+    address_identifier = create_identifier(f"{SFS}/location/address", address_hash)
+    addres_location = create_location_resource([location_type], [address_identifier], address_partOf)
+    address_entry, address_reference = create_entry_and_reference(addres_location, "Location")
+
+    location_entries.append(address_entry)
+    location_references.append(address_reference)
+
+    return location_entries, location_references
+
+
+def create_encounter(db: DatabaseSession,
+                     record: dict,
+                     patient_reference: dict,
+                     location_references: list) -> Optional[tuple]:
+    """ Returns a FHIR Encounter resource entry and reference """
+    encounter_location_references = create_encounter_location_references(db, record, location_references)
+
+    if not encounter_location_references:
+        return None, None
+
+    encounter_date = record["collection_date"]
+    # This matches how our clinical parse_uw generates encounter id
+    encounter_id = generate_hash(f"{record['mrn']}{record['accession_no']}{encounter_date}")
```

Comment:
Really good to think about matching these up! A couple of thoughts about robustness:

1. Currently `clinical parse-uw` prefers collection date, but will fallback to lab timestamp if collection date isn't available. The lab timestamp includes a time part, which would then differ from our date here and make the encounter identifiers not match. Looking at the previously imported `clinical parse-uw` dataset, it always had collection date. But I don't know if we should remove the lab timestamp fallback and require a collection date in the old format if we ever get them again to avoid identifier mismaches and duplicate encounters.

2. The MRN and accession aren't normalized by `clinical parse-uw` or this process, so casing differences will cause different identifiers. Ideally we'd normalize case and normalize various NA values to the empty string too.

Reply:
It looks like `clinical parse_uw` does `lower()` within `create_unique_identifier`! So we wouldn't have to re-do the manually pulled clinical data

Reply:
Oh I missed that completely! Nice catch.